Initial Structure and Buildout of JBI

bsieber-mozilla opened the pull request.

leplatrem commented:
How does this relate to #1 ?

grahamalama commented on src/core/config.py:
```diff
@@ -0,0 +1,60 @@
+            key, value = process(filename)
+            if key:
+                config_map[key] = value
+        except Exception as exception:  # pylint: disable=broad-except
```
Catching an `Exception` could raise an eyebrow (even the linter agrees). Are there specific errors that `process` would throw that we might want to handle differently? I do notice you have a comment at the top of the function, but maybe something closer to this line explaining why we're catching a base `Exception` would be useful.

bsieber-mozilla replied:
resolved in this commit:
https://github.com/mozilla/jira-bugzilla-integration/pull/1/commits/8219d4187d7cd8cde7dd014367fdf65aa6a9d98a

grahamalama commented on src/core/actions.py:
```diff
@@ -0,0 +1,27 @@
+
+def get_action_context_by_key(key):
+    if key not in module_dict:
+        assert False, "Unknown key requested"
```
Do you think raising a `ValueError` here might be a better choice? From the docs:
> Raised when an operation or function receives an argument that has the right type but an inappropriate value

That description seems pretty apt for this situation.

bsieber-mozilla replied:
resolved in this commit:
https://github.com/mozilla/jira-bugzilla-integration/pull/1/commits/8219d4187d7cd8cde7dd014367fdf65aa6a9d98a

grahamalama commented on src/core/config.py:
```diff
@@ -0,0 +1,60 @@
+        action = ret_dict.get("action")
+        known_actions = actions.get_action_context_by_key(key=action_key)
+        if action not in known_actions:
+            assert False, f"CONFIG ERROR: Unknown action `{action}`."
```
I noticed we do this assertion with a `CONFIG ERROR` message thing a few times in this function. Maybe this would be a good usecase for making our own `ConfigError(Exception)` and raising that instead?

bsieber-mozilla replied:
resolved in this commit:
https://github.com/mozilla/jira-bugzilla-integration/pull/1/commits/8219d4187d7cd8cde7dd014367fdf65aa6a9d98a

grahamalama reviewed (commented):
Looks like a solid start! A few inline comments, and 3 general things:

- I'm noticing that we're using assertions in places where regular exceptions might be a better choice. I'll also point out at least one or two places where I noticed it, but I think it might be a bit of a smell to use `assert` in production code. Granted, there are places where asserts can be useful, so I'm open to pushback here, but in general it's something I (personally) would avoid using

- As we kind of discussed on our call, I think some of the subpackages we're creating might not be needed just yet. Looking at it a bit more, I think all of the functions in the `core` modules could live nicely in `src.jbi.process`. I think that module has one clear job -- process the tags and actions to prepare them to be used by the FastAPI app.

- In places where you use type hints for container objects, it might be nice to further provide the type(s) for the objects in the container. E.g. `Tuple[str, str]`

grahamalama commented on src/core/config.py:
```diff
@@ -0,0 +1,60 @@
+    if not config_map:
+        config_map = {}
+
+    for filename in glob.glob(os.path.join(folder_path, "*.json")):
+        try:
+            key, value = process(filename)
```
[pathlib](https://docs.python.org/3/library/pathlib.html) might be nice here?
```suggestion
    for filename in Path(folder_path).glob("*.json"):
        try:
            key, value = process(str(filename))
```

bsieber-mozilla replied:
resolved in this commit:
https://github.com/mozilla/jira-bugzilla-integration/pull/1/commits/8219d4187d7cd8cde7dd014367fdf65aa6a9d98a

bsieber-mozilla commented:
> How does this relate to #1 ?

@leplatrem I met with graham yesterday discussing most of this PR; and will again briefly tomorrow for some general review of PR 4. 

#1 is still the primary; modifications and comments here will result in changes to #1.

This split was to reduce the amount of files viewed per-review. You can ignore #3 and #4 since it's the split history of #1 

_A future approach would be to ask for the review sooner--or split based on files/directories instead of commit history._

bsieber-mozilla commented:
> Looks like a solid start! A few inline comments, and 3 general things:
> 
>     * I'm noticing that we're using assertions in places where regular exceptions might be a better choice. I'll also point out at least one or two places where I noticed it, but I think it might be a bit of a smell to use `assert` in production code. Granted, there are places where asserts can be useful, so I'm open to pushback here, but in general it's something I (personally) would avoid using
> 
>     * As we kind of discussed on our call, I think some of the subpackages we're creating might not be needed just yet. Looking at it a bit more, I think all of the functions in the `core` modules could live nicely in `src.jbi.process`. I think that module has one clear job -- process the tags and actions to prepare them to be used by the FastAPI app.
> 
>     * In places where you use type hints for container objects, it might be nice to further provide the type(s) for the objects in the container. E.g. `Tuple[str, str]`

@grahamalama thanks for the review! I will respond to each comment individually (and absorb them into #1).

bsieber-mozilla commented on .gitignore:
```diff
@@ -127,3 +127,6 @@ dmypy.json
 .pyre/
+
+# iSort - Pre-Commit
+.isort.cfg
```
I believe this is due to linting being used in both docker and pre-commit. IIRC precommit was the culprit that needed this.

bsieber-mozilla replied:
https://github.com/mozilla/jira-bugzilla-integration/issues/8

bsieber-mozilla commented on pyproject.toml:
```diff
@@ -0,0 +1,73 @@
+'''
+
+[tool.isort]
+# https://black.readthedocs.io/en/stable/compatible_configs.html
+multi_line_output = 3
+include_trailing_comma = true
+force_grid_wrap = 0
+use_parentheses = true
+ensure_newline_before_comments = true
+line_length = 88
```
Black and iSort seem to fight via pre-commit. Whichever comes first, the other then modifies and reverses.

bsieber-mozilla replied:
https://github.com/mozilla/jira-bugzilla-integration/issues/8

bsieber-mozilla commented on src/core/config.py:
```diff
@@ -0,0 +1,60 @@
+
+
+def per_file_process(
+    filename: str, ret_dict: dict, required_keys, action_key, filename_key
```
resolved in this commit:
https://github.com/mozilla/jira-bugzilla-integration/pull/1/commits/8219d4187d7cd8cde7dd014367fdf65aa6a9d98a

bsieber-mozilla commented on pyproject.toml:
```diff
@@ -0,0 +1,73 @@
+    [tool.pylint.'MESSAGES CONTROL']
+    disable = [
+        "C0114", #missing-module-docstring
+        "C0116", #missing-function-docstring
```
https://github.com/mozilla/jira-bugzilla-integration/issues/7

Made this an issue; I believe this can be resolved over time, and would be helpful for individuals looking to contribute to it as open source.

bsieber-mozilla commented on pyproject.toml:
```diff
@@ -0,0 +1,73 @@
+    load-plugins='pylint_pytest'
+    ignore='third_party'
+
+[tool.black]
```
https://github.com/mozilla/jira-bugzilla-integration/issues/8

bsieber-mozilla commented:
> * I'm noticing that we're using assertions in places where regular exceptions might be a better choice. I'll also point out at least one or two places where I noticed it, but I think it might be a bit of a smell to use `assert` in production code. Granted, there are places where asserts can be useful, so I'm open to pushback here, but in general it's something I (personally) would avoid using

Modified to use ValueError, ActionError, ConfigError, and ProcessErrors.

> 
> * As we kind of discussed on our call, I think some of the subpackages we're creating might not be needed just yet. Looking at it a bit more, I think all of the functions in the `core` modules could live nicely in `src.jbi.process`. I think that module has one clear job -- process the tags and actions to prepare them to be used by the FastAPI app.

I'm hesitant to re-add this together; but I will do a bit more thinking on this.

> 
> * In places where you use type hints for container objects, it might be nice to further provide the type(s) for the objects in the container. E.g. `Tuple[str, str]`

Updated where this was noticed.

----

Closing this PR as most of the review has been absorbed into #1
